src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java

```diff
@@ -486,42 +488,61 @@ private void handleCurrentStateBroken(Optional<State> desiredState, State previo
     }
 
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
-    private void handleCurrentStateNew(Optional<State> desiredState) throws InterruptedException {
+    private void handleCurrentStateNewAsync(Optional<State> desiredState,
+                                       AtomicReference<Predicate<Object>> asyncFinishAction) {
         // if no desired state is set, don't do anything.
         if (!desiredState.isPresent()) {
             return;
         }
 
+        // if there is already a install() task running, do nothing.
+        Future<?> currentTask = backingTask.get();
+        if (currentTask != null && !currentTask.isDone()) {
+            return;
+        }
+
         long currentStateGeneration = stateGeneration.incrementAndGet();
+
+        Integer timeout = getTimeoutConfigValue(
+                LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
+        Future<?> schedule =
+                greengrassService.getContext().get(ScheduledExecutorService.class).schedule(() -> {
+                    if (getState().equals(State.NEW) && currentStateGeneration == getStateGeneration().get()) {
+                        greengrassService.serviceErrored(ComponentStatusCode.INSTALL_TIMEOUT, "Timeout in install");
+                    }
+                }, timeout, TimeUnit.SECONDS);
+
         replaceBackingTask(() -> {
             if (!State.NEW.equals(getState()) || getStateGeneration().get() != currentStateGeneration) {
                 // Bail out if we're not in the expected state
                 return;
             }
             try {
                 greengrassService.install();
+                if (!State.ERRORED.equals(lastReportedState.get()) && !schedule.isDone()) {
+                    internalReportState(State.INSTALLED);
+                    schedule.cancel(true);
+                }
             } catch (InterruptedException t) {
+                schedule.cancel(true);
                 logger.atWarn("service-install-interrupted").log("Service interrupted while running install");
+                Thread.currentThread().interrupt();
             } catch (Throwable t) {
+                schedule.cancel(true);
                 greengrassService.serviceErrored(t);
             }
         }, LIFECYCLE_INSTALL_NAMESPACE_TOPIC);
 
-        Integer installTimeOut = getTimeoutConfigValue(
-                LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
-
-        try {
-            backingTask.get().get(installTimeOut, TimeUnit.SECONDS);
-            if (!State.ERRORED.equals(lastReportedState.get())) {
-                internalReportState(State.INSTALLED);
+        asyncFinishAction.set((stateEvent) -> {
+            // For the first request, if the stateEvent is not a StateTransitionEvent,
+            // wait again to avoid execution twice in the installation phase.
+            if (0 == newStateUpdateCount.getAndIncrement() && stateEvent instanceof DesiredStateUpdatedEvent) {
```

Comment:
I think you should mirror `handleStateTransitionStartingToRunningAsync`:
1. if a state is reported, then it means the install() has finished, so cancel the schedule and return true;
2. if a desired state is updated, then check if the new desired state is still INSTALLED,
a. if it is, then return false
b. if it isn't, then stopBackingTask and return true

Reply:
Even though making changes like this, it still cannot solve the problem that greengrassService. install() sometimes is called twice (Has an impact on this unit test "LifecycleTest.GIVEN_state_new_WHEN_requestStart_called_THEN_install_invoked"). Is this acceptable ?

Reply:
No, this is absolutely not acceptable. Install cannot be called twice when it should only run once.